### PR TITLE
NXP: LPC11U6x: fix GPIO base DT property

### PR DESCRIPTION
**Pull request by simonguinot**
This patch series:
- fixes the gpio_lpc11u6x driver to actually makes use of the GPIO "base" property.
- fixes the GPIO "base" DT property definition following the recommendation of @galak in [#PR28818](https://github.com/zephyrproject-rtos/zephyr/pull/28818).

### CODE REVIEW

**Review by galak — CHANGES_REQUESTED** [review]
Now that I see the use of `base` we should use `gpio-reserved-ranges` for this.

https://git.kernel.org/pub/scm/linux/kernel/git/devicetree/devicetree-rebasing.git/tree/Bindings/gpio/gpio.txt#n150

We should add `gpio-reserved-ranges` to `gpio-controller.yaml ` as an optional property.

**Comment by simonguinot** [comment]
> Now that I see the use of `base` we should use `gpio-reserved-ranges` for this.
> 
> https://git.kernel.org/pub/scm/linux/kernel/git/devicetree/devicetree-rebasing.git/tree/Bindings/gpio/gpio.txt#n150
> 
> We should add `gpio-reserved-ranges` to `gpio-controller.yaml ` as an optional property.

This makes sense @galak.

Actually the following macros are used to build port_pin_mask (from which depend all the checks in GPIO core):

```
#define GPIO_PORT_PIN_MASK_FROM_NGPIOS(ngpios)                  \
        ((gpio_port_pins_t)(((uint64_t)1 << (ngpios)) - 1U))
 
#define GPIO_PORT_PIN_MASK_FROM_DT_NODE(node_id)                \
        GPIO_PORT_PIN_MASK_FROM_NGPIOS(DT_PROP(node_id, ngpios))
 
#define GPIO_PORT_PIN_MASK_FROM_DT_INST(inst)                   \
        GPIO_PORT_PIN_MASK_FROM_NGPIOS(DT_INST_PROP(inst, ngpios))
```

Unfortunately, I don't see how I could handle the `gpio-reserved-ranges` property in this macros. Indeed, `gpio-reserved-ranges` provides an array of integers sorted by tuples (e.g. `gpio-reserved-ranges = <0 4>, <12 2>;`). Honnestly I don't know what kind of DT_ macros I could use to handle that.

Obviously, I could add some C code in the GPIO driver to update port_pin_mask with `gpio-reserved-ranges` at run time. But I feel that the handling should go in the GPIO core and not in the end-driver.

Please, can you offer some guidance here ?

**Comment by galak** [comment]
@mbolivar-nordic any thoughts on how we might handle this.

1. Add support to edtlib or gen_defines.py to determine mask based on various DTS props
2. figure out some macro magic to do this in gpio.h.  We have `GENMASK` macro, just not sure how we can iterate over the array elements in `gpio-reserved-ranges` in pairs.

**Comment by mbolivar-nordic** [comment]
> @mbolivar-nordic any thoughts on how we might handle this.

Interface-vs-implementation wise, if we move `GPIO_PORT_PIN_MASK_FROM_DT_NODE(node_id)` to devicetree/gpio.h and rename it to something like `DT_GPIO_VALID_MASK(node_id)` (and similarly for a new `DT_INST_GPIO_VALID_MASK(node_id)`), then driver-side code doesn't care how we get it done. We could start with option 1. (adding support in edtlib or gen_defines) and move to macro magic later if we want to and figure out how.

**Comment by mbolivar-nordic** [comment]
@mnkp thoughts?

**Comment by mbolivar-nordic** [comment]
> @mnkp thoughts?

ping

File: drivers/gpio/gpio_lpc11u6x.c [hunk 1]
```diff
@@ -97,6 +97,7 @@ struct gpio_lpc11u6x_config {
 	const struct gpio_lpc11u6x_shared *shared;
 	char *pinmux_name;
 	uint8_t port_num;
+	uint8_t base;
```
Comment (pabigot) [inline drivers/gpio/gpio_lpc11u6x.c:100]: `base` is a little obscure.  Maybe `first-pin` or `base-pin`?

**Review by pabigot — COMMENTED** [review]
I think I agree that `gpio-reserved-ranges` generically is the right way to go; if it's not present, it should default to 
 `BITMASK(ngpios)`.

The value could be generated by synthesizing an expression like:

```
(gpio_port_pins_t)(BIT64_MASK(ngpios) \
  & ~(BIT64_MASK(hi0) & ~BIT64_MASK(lo0))  \
  ...  \
  & ~(BIT64_MASK(hiN) & ~BIT64_MASK(lo#)) \
)
```
from the devicetree property values.

**Comment by mnkp** [comment]
> Interface-vs-implementation wise, if we move `GPIO_PORT_PIN_MASK_FROM_DT_NODE(node_id)` to devicetree/gpio.h and rename it to something like `DT_GPIO_VALID_MASK(node_id)` (and similarly for a new `DT_INST_GPIO_VALID_MASK(node_id)`), then driver-side code doesn't care how we get it done. We could start with option 1. (adding support in edtlib or gen_defines) and move to macro magic later if we want to and figure out how.

We use `GPIO_PORT_PIN_MASK_FROM_DT_NODE` as the name of the macro since we need to fill in `port_pin_mask` field in `struct gpio_driver_config` defined in `include/drivers/gpio.h` that is common to all GPIO drivers. 
```
struct gpio_driver_config {
	/* Mask identifying pins supported by the controller.
	 *
	 * Initialization of this mask is the responsibility of device
	 * instance generation in the driver.
	 */
	gpio_port_pins_t port_pin_mask;
};
```
The proposed new name of the macro `DT_GPIO_VALID_MASK` looks indeed better, its meaning is clearer. If we rename the macro we should also rename the struct field.

That said I'm not sure I fully understood the proposal from @mbolivar-nordic since already today the driver-side code doesn't care how we get it done. The only problem is that there is no easy way to implement the macro that takes into account `ngpios` and `gpio-reserved-ranges` DTS properties. That's why @galak proposed
```
1. Add support to edtlib or gen_defines.py to determine mask based on various DTS props
```

Also from my point of view using standard `gpio-reserved-ranges` DTS property would be much preferable than non-standard `base`.

**Comment by mbolivar-nordic** [comment]
@mnkp I am in agreement that we can add special-purpose support in gen_defines.py to start with as @galak says, and then if we figure out some fancy way to do it in the preprocessor instead using only existing API s, we can migrate it to devicetree/gpio.h. That's what I was trying to say in this comment: https://github.com/zephyrproject-rtos/zephyr/pull/28877#issuecomment-704352987

So I think we're all on the same page for this approach.

**Pull request closed without merging**